runtime/compiler/env/DependencyTable.cpp

```diff
@@ -60,17 +178,42 @@ TR_AOTDependencyTable::classLoadEvent(TR_OpaqueClassBlock *clazz, bool isClassLo
    catch (std::exception&)
       {
       deactivateTable();
+      return;
+      }
+
+   resolvePendingLoads();
+   }
+
+void
+TR_AOTDependencyTable::checkForSatisfaction(PersistentUnorderedSet<MethodEntryRef> waitingMethods)
+   {
+   for (auto &entry: waitingMethods)
+      {
+      if (entry->second._remainingDependencies)
+         _pendingLoads.insert(entry);
+      else
+         --entry->second._remainingDependencies;
       }
    }
 
 void
 TR_AOTDependencyTable::classLoadEventAtOffset(J9Class *ramClass, uintptr_t offset, bool isClassLoad, bool isClassInitialization)
    {
-   auto entry = getOffsetEntry(offset, isClassLoad);
-   TR_ASSERT(entry || !isClassLoad, "Class %p offset %lu initialized without loading");
+   auto offsetEntry = getOffsetEntry(offset, isClassLoad);
+   TR_ASSERT(offsetEntry || !isClassLoad, "Class %p offset %lu initialized without loading", ramClass, offset);
+
+   // Check for dependency satisfaction if this is the first class initialized
+   // for this offset
+   if (isClassInitialization && (NULL != findCandidateForDependency(offsetEntry->_loadedClasses, true)))
+      checkForSatisfaction(offsetEntry->_waitingInitMethods);
 
+   // Track the class, and also check for dependency satisfaction if this is the
+   // first class loaded for this offset
    if (isClassLoad)
-      entry->_loadedClasses.insert(ramClass);
+      {
```

Comment:
I think this bit of the code was confusing because in my refactoring I accidentally reversed a couple of the conditions. Hopefully it's clearer now.

If a class is being loaded, then we want to update the methods waiting for load if there aren't any previous loads for this offset, and if a class is being initialized, then we want to update the dependencies waiting for init if there aren't any initialized classes for this offset yet. The order doesn't end up mattering, because `ramClass` itself should never be considered an already-existing candidate.

Reply:
Will be clearer, that is, once I push the corrections